### DISCUSSION TITLE
Integration & configuration with/for Blocky (and/or other ad-blocker like pihole)

### DISCUSSION THREAD

**Discussion by Parrat0** [BeryJu/gravity#1336] [General]
Hello!

I'm using gravity in my homelab primarily as an authoritative dns-server with corresponding DHCP server across multiple subnets, even clustered. However I am confused about the Blocky integration for ad-blocking and how to set this up.

Currently for ad-blocking I have pihole which I forward all the dns requests to from gravity, but in pihole I only see gravities ip-address.  Initially I was looking for a way to forward a dns request with source ip to pihole, is this possible?

Second I also want to move to Blocky instead of pihole so I can have both dns/dhcp/ad-blocking clustered. However I'm confused on how and where to configure Blocky:

- Do I need to first create a blocky docker and then forward the dns request?
- do I then still need to do the integration you describe here: https://gravity.beryju.io/docs/dns/zones/#forward_blocky? 
- or is this in integration creating the Blocky config for me?

Cheers!

**Comment by BeryJu** [comment]
I dont think it's easily possible to forward DNS requests while preserving the client IP. It would in theory be possible using the [Proxy protocol](https://www.haproxy.org/download/1.8/doc/proxy-protocol.txt), which would require both gravity to support sending requests with it and pihole/blocky to receive requests with it.

The blocky zone handler runs blocky as a library within gravity, so you don't need a separate blocky container. This also means that you get correct logs and metrics with client IPs etc. You can simply add the blocky handler as is in the documentation and it'll enable ad-blocking on the specified zone.

**Comment by Parrat0** [comment]
Right! Thank you for the help, it took some time to experiment but I got it working.

I ended up setting up the root zone as in your documentation under "forward blocky":
![image](https://github.com/user-attachments/assets/0de0eaf0-fab9-4d1c-8f5c-d660d69ba696)

A couple of notes though that I've run into:
I have to adhere to a specific config syntax and tab/space usage otherwise it's not accepted by Gravity. This is logical and it's in the documentation but the error message I get otherwise is super cryptic. So RTFM(!)
![image](https://github.com/user-attachments/assets/c44bd18f-8b18-4cc1-ab33-f0ba4a7ad715)

Then after the full config gets accepted by Gravity, the config file itself gets a red color and gets reformatted:
- color changes to red
- "type: forward_blocky" get put at the very end of the file
- the actual blocky config text is put first

Here is a screenshot of how mine looks like:
![image](https://github.com/user-attachments/assets/4e0c9618-2847-44d2-a3fc-930a39b1431b)

Now the strangest thing happens: the very first line get reformatted into:
` - config: >`

Which is fine until you want to change the config, because changing the config in its entirety with this line present will crash Gravity every time. For example I want to add a whitelist url to Blocky config: 

1. I would copy the current config into some editor,
2. add the whitelist url lines,
3. copy it back to Gravity. 

Gravity accepts and saves the config and then promptly crashes. 
I cannot change the config in Gravity itself while relying on Blocky because I can never get the json correct. Gravity gives a very cryptic error message about json not being valid.

So to edit the config I have to go back to the original one I used, so the one with the proper formatting based on the documentation and edit that one. Then copy/paste into Gravity so the config looks again like the documentation, and my screenshot:
![image](https://github.com/user-attachments/assets/76b9538c-e69f-49a0-9b58-af17cf181f5a)
This way is the only way I can update the config.

So two questions: 
1 - is this working as intended?
2 - is it possible to reference the blocky config file in some way? So for example the config on Gravity would look like this:
![image](https://github.com/user-attachments/assets/b8f5ae08-0218-4b93-96f5-57588573218b)

**Reply by BeryJu** [reply]
interesting, having an invalid config should definitely not crash gravity. I would also add that in most cases using the custom config is not required as you can define blocklists and allowlists directly without having to supply a custom config

the re-formatting happens because yaml, and because the blocky config is a yaml file as a string within another yaml file.

importing from file probably won't be added as the point of gravity is to store all required config in etcd so it's distributed

**Reply by Parrat0** [reply]
ahh wait now I understand. So if I use the "Full custom config" I'm loading an external config into the Gravity blocky instance and overriding the default that is comes with. Whereas if I use the customization options under the "forward_blocky" header here [https://gravity.beryju.io/docs/dns/zones/](url); I'll be using the built in config. Ok this didn't click until now. 

I'm making it too difficult for myself, I'll redo the root domain config.

Thank you!
